File: 35_reverse-linked-list/reverse-linked-list.cpp
/*
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/reverse-linked-list
@Language: C++
@Datetime: 16-08-04 02:08
*/

/**
 * Definition of ListNode
 * 
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 * 
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: The new head of reversed linked list.
     */
    ListNode *reverse(ListNode *head) {
        // write your code here
        ListNode * dummy = new ListNode(0);
        ListNode * p;
        while (head) {
            p = head;
            head = head->next;
            p->next = dummy->next;
            dummy->next = p;
        }
        p = dummy->next;
        delete dummy;
        return p;
    }
};

File: 35_reverse-linked-list/reverse-linked-list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/reverse-linked-list
@Language: Python
@Datetime: 16-06-10 10:06
'''

"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: You should return the head of the reversed linked list. 
                  Reverse it in-place.
    """
    def reverse(self, head):
        # write your code here
        if head is None or head.next is None:  # only one node
            return head
        prev, next = head, head.next
        while next:
            theNext = next.next
            next.next = prev
            prev = next
            next = theNext
        head.next = None
        return prev